Add VIC-20 support

elderling opened the pull request:
Add conditional assembly based on target
Add Vic-20 ROM calls
Add VIC-20 example linker .cfg for 32K VIC-20 ONLY
Add separate makefile for VIC-20

elderling committed:
Add VIC-20 support

Add conditional assembly based on target
Add Vic-20 ROM calls
Add VIC-20 example linker .cfg for 32K VIC-20 ONLY
Add separate makefile for VIC-20

mrdudz commented on float.s:
```diff
@@ -12,22 +12,44 @@ BINARYFORMAT = BINARYFORMAT_CBM_PACKED
         cli
         rts
+.endif
+
```
Is there a particular reason for not moving the __basicon and __basicoff symbols into the .if statement?

elderling replied:
I suspected that the assembler might choke on conditionally declared symbols. I didn't check it both ways. I have no objection to changing it.

mrdudz commented on float.s:
```diff
@@ -74,15 +96,15 @@ ___float_u16_to_fac:
-        jsr __basicon
+        __enable_basic_if_needed
         jsr BASIC_FAC_to_u16
         jsr __basicoff
```
please also substitute the __basicoff calls by a macro (__disable_basic_if_needed)

elderling replied:
Will do.

elderling replied:
OK, I think I originally misread this. I already have a macro `__return_with_cleanup` that I should have used here. I gave it an "asymmetric" because `__disable_basic_if_needed` doesn't make it clear that the macro results in a change to execution flow.

mrdudz commented on float.s:
```diff
@@ -491,21 +513,21 @@ __float_ret2:
-        stx $01
-        cli
+        ;stx $01
+        ;cli
```
are you sure this is correct?

elderling replied:
This appears to be a flat out mistake. Will address.

mrdudz commented on vic20-32k.cfg:
```diff
@@ -0,0 +1,42 @@
+# Memory configuration for the VIC-20 with 32K RAM Cartridge
```
do we really need the vic20 linker config here? isnt there one in the cc65 tree?

elderling replied:
Well, I think of it more like "a" VIC-20 linker config than "the" linker config. The default config is for unexpanded VIC-20 and the test program won't fit into an unexpanded VIC-20's memory. Is there a straightforward, system-agnostic way to point to a linker config that's part of the cc65 tree? I'm only aware of the -t flag for the target and the -C flag for a specific file. I suppose it's also worth mentioning that I switch back and forth between Windows & Linux and I'm pretty sure I've compiled my own cc65 binary on Linux (as opposed to using something from a package manager).

mrdudz replied:
just -C vic20-32k.cfg should do the trick when that file is present in the cfg dir in the cc65 tree

elderling replied:
OK, file removed!

mrdudz commented:
Am Dienstag, 28. September 2021, 18:39:44 CEST schrieb Jonathan W. Taylor:
> @elderling commented on this pull request.
> 
> 
> 
> 
>  ; return to C, FAC as unsigned int
>  __float_fac_to_u16:
> -        jsr __basicon
> +        __enable_basic_if_needed
>          jsr BASIC_FAC_to_u16
>          jsr __basicoff
> 
> OK, I think I originally misread this. I already have a macro
> `__return_with_cleanup` that I should have used here. I gave it an
> "asymmetric" because `__disable_basic_if_needed` doesn't make it clear that
> the macro results in a change to execution flow.

there are places where __basicoff was called with jmp (which is where your 
macro can be used) and others where its called by jsr - which need another 
macro that ends with rts

-- 

http://hitmen.eu                 http://ar.pokefinder.org
http://vice-emu.sourceforge.net  http://magicdisk.untergrund.net

Wenn Sie schon Raubkopien machen, dann stellen Sie wenigstens sicher, dass die 
Qualitaet gut ist. [...] Zeigen Sie etwas Respekt für das Gut, das Sie 
stehlen. 
<Terry Gilliam>

elderling committed:
Add __disable_basic_if_needed macro

elderling committed:
Replace commented code with proper .if conditional

elderling committed:
Add explicit -t c64 to makefile for float.s

It is my understanding that c64 is supposed to be the default, but
sometimes the float.s would assemble as if __C64__ weren't defined.
Don't know if this is a bug in ca65 or not but specifying the target on
the command line via -t always produces correct results.

elderling committed:
Simplify conditional assembly of c64-only macros

elderling committed:
Remove onneeded VIC-20 linker .cfg

cl65 looks in its own cfg directory for files specified by
-C

elderling commented:
What do I need to do next? I'm willing to re-squash this and do a separate pull request if you'd prefer? This is the first time I've done a pull request that wasn't automatically merged in github so I appreciate your patience ...

mrdudz commented:
> What do I need to do next? I'm willing to re-squash this and do a separate pull request if you'd prefer? This is the first time I've done a pull request that wasn't automatically merged in github so I appreciate your patience ...

Sorry, i was busy with other stuff.... Have you tested and checked if it still works for C64? If so, it looks good and can be merged :)

elderling commented:
Works for C64 as far as I can tell. I even tried uncommenting a bunch of stuff in `floattest.c` and didn't see any problems.

mrdudz commented:
so, merged. cheers!
